clients/src/main/java/org/apache/kafka/common/record/MemoryRecords.java

```diff
@@ -363,57 +377,89 @@ public static MemoryRecordsBuilder builder(ByteBuffer buffer,
+                RecordBatch.NO_PRODUCER_EPOCH, RecordBatch.NO_SEQUENCE, RecordBatch.NO_PARTITION_LEADER_EPOCH,
+                false, records);
     }
 
-    private static MemoryRecords withRecords(byte magic, long initialOffset, CompressionType compressionType,
-                                             TimestampType timestampType, long producerId, short producerEpoch,
-                                             int baseSequence, int partitionLeaderEpoch, SimpleRecord ... records) {
+    public static MemoryRecords withRecords(byte magic, long initialOffset, CompressionType compressionType,
+                                            TimestampType timestampType, long producerId, short producerEpoch,
+                                            int baseSequence, int partitionLeaderEpoch, boolean isTransactional,
+                                            SimpleRecord ... records) {
         if (records.length == 0)
             return MemoryRecords.EMPTY;
         int sizeEstimate = AbstractRecords.estimateSizeInBytes(magic, compressionType, Arrays.asList(records));
         ByteBuffer buffer = ByteBuffer.allocate(sizeEstimate);
         long logAppendTime = RecordBatch.NO_TIMESTAMP;
         if (timestampType == TimestampType.LOG_APPEND_TIME)
             logAppendTime = System.currentTimeMillis();
-        MemoryRecordsBuilder builder = builder(buffer, magic, compressionType, timestampType, initialOffset,
-                logAppendTime, producerId, producerEpoch, baseSequence, false, partitionLeaderEpoch);
+        MemoryRecordsBuilder builder = new MemoryRecordsBuilder(buffer, magic, compressionType, timestampType,
+                initialOffset, logAppendTime, producerId, producerEpoch, baseSequence, isTransactional,
+                partitionLeaderEpoch, buffer.remaining());
         for (SimpleRecord record : records)
             builder.append(record);
         return builder.build();
     }
 
+    public static MemoryRecords withEndTransactionMarker(EndTransactionMarker marker, long producerId, short producerEpoch) {
+        return withEndTransactionMarker(0L, marker, producerId, producerEpoch);
+    }
+
+    public static MemoryRecords withEndTransactionMarker(long initialOffset, EndTransactionMarker marker,
+                                                         long producerId, short producerEpoch) {
+        ByteBuffer buffer = ByteBuffer.allocate(128);
+        MemoryRecordsBuilder builder = builder(buffer, RecordBatch.CURRENT_MAGIC_VALUE, CompressionType.NONE,
+                TimestampType.CREATE_TIME, initialOffset, RecordBatch.NO_TIMESTAMP, producerId, producerEpoch,
+                RecordBatch.CONTROL_SEQUENCE, true, RecordBatch.NO_PARTITION_LEADER_EPOCH);
+        ByteBuffer value = marker.serializeValue();
+        builder.appendControlRecord(System.currentTimeMillis(), marker.controlType(), value);
```

Comment:
I was trying to keep `appendControlRecord` generic for future control record types. That said, it seems reasonable to add an `appendEndTxnMarker` which handles serialization of both the key and value.